omp/multigrid/amgx_pgm_kernels.cpp

```diff
@@ -62,22 +68,62 @@ namespace amgx_pgm {
 template <typename IndexType>
 void match_edge(std::shared_ptr<const OmpExecutor> exec,
                 const Array<IndexType> &strongest_neighbor,
-                Array<IndexType> &agg) GKO_NOT_IMPLEMENTED;
+                Array<IndexType> &agg)
+{
+    auto agg_vals = agg.get_data();
+    auto strongest_neighbor_vals = strongest_neighbor.get_const_data();
+#pragma omp parallel for
+    for (size_type i = 0; i < agg.get_num_elems(); i++) {
+        if (agg_vals[i] == -1) {
+            size_type neighbor = strongest_neighbor_vals[i];
+            if (neighbor != -1 && strongest_neighbor_vals[neighbor] == i) {
+                // Use the smaller index as agg point
+                auto group = min(i, neighbor);
+                agg_vals[i] = group;
+                agg_vals[neighbor] = group;
```

Comment:
it should be okay on the results aspect.
they always set the smaller index, so they should write the same result from two points.
I can do it with condition but I need to check what impact is.

Reply:
It depends entirely on what the neighbors set look like. If it's symmetric (neighbor[i] = j and neighbor[j] = i), then there should be no issue. But if you have cases where neighbor[i] = j, neighbor[j] = k and neighbor[k] = i, then you could get different results depending on what gets executed first, I think?

Reply:
This case you mentioned does not work in the previous one, either.
Those points with circular strong neighbor will be assigned in the last step.
The original one: the both side of strongest connection will set the smaller index as aggregation group.
Current one: only the smaller one proceed settin the aggregation of both side.

Reply:
I think especially considering the weaker memory model for C++ on ARM processors, we should give this some thought. There is a potential RAW race condition between the write in  `agg_vals[neighbor] = i;` and the read in `agg_vals[i] == -1`. This might be especially problematic if the reads and writes are torn, i.e. implemented as two distinct memory transactions, and potentially even reordered.
If any write to `agg_vals[neighbor]` become visible to the thread where `i == neighbor`, the comparison will become false. I would argue though that this cannot cause any issues, since in the writing thread, `i < neighbor`, so the reading thread would have `i > neighbor` and thus would not write to memory itself. Does that sound right to you?

This will definitely produce ThreadSanitizer warnings, though, not sure how to silence those.

Reply:
Sorry for late reply.
Yes, read may be after or before the write from the smaller threads and it will not cause any issue as the reason you said.
I am not sure how ThreadSanitizer warning shows on this part.